File: tarot-game/Assets/Scripts/WebSocketMessages.cs
using System.Collections.Generic;

[System.Serializable]
public class WebSocketMessage {
    public string type;
}

[System.Serializable]
public class PlayerJoined : WebSocketMessage {
    public List<string> users;
}

[System.Serializable]
public class PlayerLeft : WebSocketMessage {
    public string username;

    public PlayerLeft(string username)
    {
        this.type = "player_left";
        this.username = username;
    }
}

[System.Serializable]
public class CardPlayed : WebSocketMessage {
    public int card;
    public string username = "";

    public CardPlayed(int card)
    {
        this.type = "card_played";
        this.card = card;
    }
}

[System.Serializable]
public class HandDealt : WebSocketMessage
{
    public List<int> cards;
}

[System.Serializable]
public class StartGame : WebSocketMessage
{
    public StartGame()
    {
        this.type = "start_game";
    }
}

[System.Serializable]
public class GameReady : WebSocketMessage
{
    public GameReady()
    {
        this.type = "game_ready";
    }
}

[System.Serializable]
public class PlayerTurn : WebSocketMessage
{
    public string username;
}

[System.Serializable]
public class UsernameWrapper : WebSocketMessage
{
    public string username;
}

[System.Serializable]
public class BidMade : WebSocketMessage
{
    public string bid;
    public string username = "";

    public BidMade(Bid bid)
    {
        this.type = "bid_made";
        this.bid = bid.ToString();
    }
}

[System.Serializable]
public class BidWon : WebSocketMessage
{
    public Bid bid;
    public string username;
}

[System.Serializable]
public class TurnWon : WebSocketMessage
{
    public string username;
}

[System.Serializable]
public class GameOver : WebSocketMessage
{
    public bool victory;
    public int score;
    public List<PlayerScore> playerScores;
}

[System.Serializable]
public class PlayerScore
{
    public string username;
    public int score;
}

[System.Serializable]
public class DogReveal : WebSocketMessage
{
    public List<int> cards;
    public string attacker;
}

[System.Serializable]
public class DogMade : WebSocketMessage
{
    public List<int> cards;

    public DogMade(List<int> cards)
    {
        this.type = "dog_made";
        this.cards = cards;
    }
}

[System.Serializable]
public class AwaitAppel : WebSocketMessage
{
    public string username;
}

[System.Serializable]
public class Appel : WebSocketMessage
{
    public TarotColor color;
    public string username;

    public Appel(TarotColor color, string username)
    {
        this.color = color;
        this.username = username;
    }
}

File: tarot-backend/src/main/kotlin/data/server/WebSocketMessages.kt
package xclavel.data.server

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import xclavel.data.tarot.Color
import xclavel.data.tarot.data.tarot.Bid

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("type")
sealed class WebSocketMessage

@Serializable
@SerialName("player_joined")
data class PlayerJoined(val users: List<String>) : WebSocketMessage()

@Serializable
@SerialName("player_left")
data class PlayerLeft(val username: String) : WebSocketMessage()

@Serializable
@SerialName("player_turn")
data class PlayerTurn(val username: String) : WebSocketMessage()

@Serializable
@SerialName("card_played")
data class CardPlayed(val card: Int, val username: String) : WebSocketMessage()

@Serializable
@SerialName("turn_won")
data class TurnWon(val username: String) : WebSocketMessage()


@Serializable
@SerialName("dog_reveal")
data class DogReveal(val attacker: String, val cards: List<Int>) : WebSocketMessage()

@Serializable
@SerialName("dog_made")
data class DogMade(val cards: List<Int>): WebSocketMessage()

//Sent by client to declare intent to send game
//Sent by server to broadcast game start
@Serializable
@SerialName("start_game")
class StartGame() : WebSocketMessage()

//Sent by client to signal game scene is fully loaded
@Serializable
@SerialName("game_ready")
class GameReady(): WebSocketMessage()

@Serializable
@SerialName("game_over")
data class GameOver(val victory: Boolean, val score: Int, val playerScores: List<PlayerScore>): WebSocketMessage()

@Serializable
data class PlayerScore(val username: String, val score: Int)

@Serializable
@SerialName("hand_dealt")
data class HandDealt(val cards: List<Int>): WebSocketMessage()

//Bids

@Serializable
@SerialName("await_bid")
data class AwaitBid(val username: String): WebSocketMessage()

@Serializable
@SerialName("bid_made")
data class BidMade(val bid: Bid, val username: String) : WebSocketMessage()

@Serializable
@SerialName("bid_won")
data class BidWon(val username: String, val bid: Bid): WebSocketMessage()

@Serializable
@SerialName("fausse_donne")
class FausseDonne : WebSocketMessage()

//Appel

@Serializable
@SerialName("await_appel")
data class AwaitAppel(val username: String): WebSocketMessage()

@Serializable
@SerialName("appel")
data class Appel(val color: Color, val username: String): WebSocketMessage()

@Serializable
@SerialName("first_turn")
data class FirstTurn(val username: String): WebSocketMessage()